File: Python/medium_truck.py
from Truck import Truck
from abc import ABC

class MediumTruck(Truck, ABC):
    MAX_VOLUME = 2000
    MAX_WEIGHT = 4000

    def add_pack(self, pack):
        if self.total_volume() + pack.get_volume() <= MediumTruck.MAX_VOLUME and \
                self.total_weight() + pack.get_weight() <= MediumTruck.MAX_WEIGHT:
            self.packages.append(pack)
        else:
            return False

    def total_volume(self):
        total = 0
        for pack in self.packages:
            total += pack.get_volume()
        return total

    def total_weight(self):
        total = 0
        for pack in self.packages:
            total += pack.get_weight()
        return total

File: CPP/MediumTruck.cpp
#include "headers/MediumTruck.h"

MediumTruck::MediumTruck() {
    truckID = count;
}

int MediumTruck::getMaxWeight() const {
    return MAX_WEIGHT;
}

int MediumTruck::getMaxVolume() const {
    return MAX_VOLUME;
}

void MediumTruck::addPack(const Pack &p) {
    if (totalPacksVolume() + p.getVolume() <= MAX_VOLUME &&
        totalPacksWeight() + p.getWeight() <= MAX_WEIGHT) {
        packages.add(p);
    }
}

int MediumTruck::totalPacksVolume() {
    int total = 0;
    for (int i = 0; i < packages.size(); i++) {
        total += packages.get(i).getVolume();
    }
    return total;
}

int MediumTruck::totalPacksWeight() {
    int total = 0;
    for (int i = 0; i < packages.size(); i++) {
        total += packages.get(i).getWeight();
    }
    return total;
}
